### PR TITLE
Namespace file loader

### PR DESCRIPTION
**Pull request by defint**
@ilteoood 
Seems like test workflow has been failed, but local runs are ok.
Can you help me with this?

### CODE REVIEW

File: lib/loaders/namespace_file_translation_loader.dart [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+  NamespaceFileTranslationLoader(
+      {@required this.namespaces,
+      this.fallbackDir = "en",
+      this.basePath = "assets/flutter_i18n",
+      this.useCountryCode = false,
+      this.forcedLocale}) {
+    assetBundle = rootBundle;
+
+    assert(namespaces != null);
```
Comment (ilteoood) [inline lib/loaders/namespace_file_translation_loader.dart]: I think that is cleaner to put this checks as first statements

File: lib/loaders/namespace_file_translation_loader.dart [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    assert(namespaces.length > 0);
+  }
+
+  Future<Map> load() async {
```
Comment (ilteoood) [inline lib/loaders/namespace_file_translation_loader.dart]: most of this file is equal to the FileTranslationLoader.
I suggest to modify the visibility of the methods of that class, from private to proctected, in order o reduce the amount of code.

**Comment by ilteoood** [comment]
Hi @defint,
in your case is failing because you haven't configured the PUB_CREDENTIALS secret in your fork.
You should insert, inside that secret, the content of
~/.pub-cache/credentials.json

File: lib/loaders/namespace_file_translation_loader.dart (hunk 1 above)
Comment (defint) [reply lib/loaders/namespace_file_translation_loader.dart]: Fixed

File: lib/loaders/namespace_file_translation_loader.dart (hunk 2 above)
Comment (defint) [reply lib/loaders/namespace_file_translation_loader.dart]: Made some changes. Please check.

**Comment by defint** [comment]
@ilteoood 
Can you please invite me to be a contributor? Seems like with fork flow is not easy to deliver new features.

**Comment by ilteoood** [comment]
@defint 
Here you go

File: lib/loaders/file_translation_loader.dart [hunk 3]
```diff
@@ -50,41 +50,45 @@ class FileTranslationLoader extends TranslationLoader {
...
     }
   }
 
-  Future<void> _loadFile(final String fileName) async {
+  Future<Map> loadFile(final String fileName) async {
     try {
-      await _decodeFile(fileName, 'json', json.decode);
+      var data = await _decodeFile(fileName, 'json', json.decode);
       MessagePrinter.info("JSON file loaded for $fileName");
+      return data;
```
Comment (ilteoood) [inline lib/loaders/file_translation_loader.dart]: I prefer a single point of return

File: lib/loaders/file_translation_loader.dart [hunk 4]
```diff
@@ -50,41 +50,45 @@ class FileTranslationLoader extends TranslationLoader {
...
+    _decodedMap = await loadFile(composeFileName());
   }
 
   Future _loadFallback() async {
     try {
-      await _loadFile(fallbackFile);
+      _decodedMap = await loadFile(fallbackFile);
     } catch (e) {
       MessagePrinter.debug('Error loading translation fallback $e');
       _decodedMap = Map();
```
Comment (ilteoood) [inline lib/loaders/file_translation_loader.dart]: at this point, _decodedMap can be declared empty

File: lib/loaders/namespace_file_translation_loader.dart [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    await _loadTranslation();
+    return _decodedMap;
+  }
+
+  Future<void> _loadTranslation() async {
+    this.locale = locale ?? await findCurrentLocale();
+    MessagePrinter.info("The current locale is ${this.locale}");
+
+    for (var namespace in namespaces) {
+      try {
```
Comment (ilteoood) [inline lib/loaders/namespace_file_translation_loader.dart]: what if we put the body of this for statement in a Future function, and then use Future.wait to speed up the map loading?

File: lib/loaders/file_translation_loader.dart (hunk 3 above)
Comment (defint) [reply lib/loaders/file_translation_loader.dart]: Done

File: lib/loaders/namespace_file_translation_loader.dart (hunk 5 above)
Comment (defint) [reply lib/loaders/namespace_file_translation_loader.dart]: Good point

File: lib/loaders/file_translation_loader.dart (hunk 4 above)
Comment (defint) [reply lib/loaders/file_translation_loader.dart]: Move definition to declaration.

File: lib/loaders/namespace_file_translation_loader.dart [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    assert(namespaces.length > 0);
+
+    assetBundle = rootBundle;
+  }
+
+  Future<Map> load() async {
+    this.locale = locale ?? await findCurrentLocale();
+    MessagePrinter.info("The current locale is ${this.locale}");
+
+    List<Future<void>> waitList =
```
Comment (ilteoood) [inline lib/loaders/namespace_file_translation_loader.dart]: is the list necessary?
AFAIK Future.wait works also with arrays

Comment (defint) [reply lib/loaders/namespace_file_translation_loader.dart]: But array and list are the same things in the Dart. What do you mean? Can you please provide code example?

Comment (ilteoood) [reply lib/loaders/namespace_file_translation_loader.dart]: I mean, the .map function already produce an array, that is an iterable that you can pass to Futute.wait.
The usage of. toList can be avoided IMO

Comment (defint) [reply lib/loaders/namespace_file_translation_loader.dart]: You are right! Fixed.

**Pull request merged**